nuxeo-features/nuxeo-platform-oauth/src/main/java/org/nuxeo/ecm/platform/oauth2/request/AuthorizationRequest.java

```diff
@@ -129,22 +132,30 @@ protected AuthorizationRequest(Map<String, Serializable> map) {
 
     public OAuth2Error checkError() {
         // Check mandatory fields
-        if (StringUtils.isBlank(responseType) || StringUtils.isBlank(clientId) || StringUtils.isBlank(redirectUri)
-                || !OAuth2Client.isRedirectURIValid(redirectUri) || StringUtils.isBlank(state)) {
-            return OAuth2Error.INVALID_REQUEST;
+        if (StringUtils.isBlank(clientId)) {
+            return OAuth2Error.invalidRequest(String.format(MISSING_REQUIRED_FIELD_MESSAGE, CLIENT_ID_PARAM));
+        }
+        if (StringUtils.isBlank(responseType)) {
+            return OAuth2Error.invalidRequest(String.format(MISSING_REQUIRED_FIELD_MESSAGE, RESPONSE_TYPE_PARAM));
+        }
+        if (StringUtils.isBlank(redirectUri)) {
+            return OAuth2Error.invalidRequest(String.format(MISSING_REQUIRED_FIELD_MESSAGE, REDIRECT_URI_PARAM));
+        }
+        if (StringUtils.isBlank(state)) {
+            return OAuth2Error.invalidRequest(String.format(MISSING_REQUIRED_FIELD_MESSAGE, STATE_PARAM));
         }
 
         // Check if client exists
         ClientRegistry registry = Framework.getService(ClientRegistry.class);
-        OAuth2Client client = null;
+        OAuth2Client client;
         try {
             client = registry.getClient(clientId);
             if (client == null || !client.isEnabled()) {
-                return OAuth2Error.UNAUTHORIZED_CLIENT;
+                return OAuth2Error.unauthorizedClient(String.format("Invalid %s: %s.", CLIENT_ID_PARAM, clientId));
             }
         } catch (DirectoryException e) {
             log.warn(e, e);
-            return OAuth2Error.SERVER_ERROR;
+            return OAuth2Error.serverError("Server Error");
```

Comment:
In development mode, could it be interesting to also return a part of stacktrace ?